sygnal/webpushpushkin.py

```diff
@@ -0,0 +1,167 @@
+
+from .exceptions import PushkinSetupException
+from .notifications import ConcurrencyLimitedPushkin
+
+logger = logging.getLogger(__name__)
+
+MAX_TRIES = 3
+RETRY_DELAY_BASE = 10
+MAX_BYTES_PER_FIELD = 1024
+
+DEFAULT_MAX_CONNECTIONS = 20
+
+
+class WebpushPushkin(ConcurrencyLimitedPushkin):
+    """
+    Pushkin that relays notifications to Google/Firebase Cloud Messaging.
+    """
+
+    UNDERSTOOD_CONFIG_FIELDS = {
+        "type",
+        "max_connections",
+        "vapid_private_key",
+        "vapid_contact_email",
+    } | ConcurrencyLimitedPushkin.UNDERSTOOD_CONFIG_FIELDS
+
+    def __init__(self, name, sygnal, config):
+        super(WebpushPushkin, self).__init__(name, sygnal, config)
+
+        nonunderstood = set(self.cfg.keys()).difference(self.UNDERSTOOD_CONFIG_FIELDS)
+        if len(nonunderstood) > 0:
+            logger.warning(
+                "The following configuration fields are not understood: %s",
+                nonunderstood,
+            )
+
+        self.http_pool = HTTPConnectionPool(reactor=sygnal.reactor)
+        self.max_connections = self.get_config(
+            "max_connections", DEFAULT_MAX_CONNECTIONS
+        )
+        self.connection_semaphore = DeferredSemaphore(self.max_connections)
+        self.http_pool.maxPersistentPerHost = self.max_connections
+
+        tls_client_options_factory = ClientTLSOptionsFactory()
+
+        # use the Sygnal global proxy configuration
+        proxy_url = sygnal.config.get("proxy")
+
+        self.http_agent = ProxyAgent(
+            reactor=sygnal.reactor,
+            pool=self.http_pool,
+            contextFactory=tls_client_options_factory,
+            proxy_url_str=proxy_url,
+        )
+        self.http_agent_wrapper = HttpAgentWrapper(self.http_agent)
+
+        privkey_filename = self.get_config("vapid_private_key")
+        self.vapid_private_key = Vapid.from_file(private_key_file=self.get_config("vapid_private_key"))
+        vapid_contact_email = self.get_config("vapid_contact_email")
+        self.vapid_claims = {"sub": "mailto:{}".format(vapid_contact_email)}
+
+    async def _dispatch_notification_unlimited(self, n, device, context):
+        p256dh = device.pushkey
+        endpoint = device.data["endpoint"]
+        session_id = device.data["session_id"]
+        auth = device.data["auth"]
+        subscription_info = {
+            'endpoint': endpoint,
+            'keys': {
+                'p256dh': p256dh,
+                'auth': auth
+            }
+        }
+        payload = {
+            'room_name': n.room_name,
+            'room_alias': n.room_alias,
+            'prio': n.prio,
+            'membership': n.membership,
+            'sender_display_name': n.sender_display_name,
+            'event_id': n.event_id,
+            'room_id': n.room_id,
+            'user_is_target': n.user_is_target,
+            'type': n.type,
+            'sender': n.sender,
+            'session_id': session_id,
+        }
```

Comment:
What happens if some of these values are missing from `n`?

Reply:
Yeah, still need to work on the format of the actual payload :+1: